sigopt/xgboost/run.py

```diff
@@ -24,17 +27,69 @@ def parse_run_options(run_options):
   return run_options_parsed
 
 
+class SigOptCheckpointCallback(xgboost.callback.TrainingCallback):
+  def __init__(self, run, period=1):
+    self.run = run
+    self.period = period
+    self._latest = None
+    super().__init__()
+
+  def after_iteration(self, model, epoch, evals_log):
+    if not evals_log:
+      return False
+
+    checkpoint_logs = {}
+    for data, metric in evals_log.items():
+      for metric_name, log in metric.items():
+        if isinstance(log[-1], tuple):
+          score = log[-1][0]
+        else:
+          score = log[-1]
+        checkpoint_logs.update({'-'.join((data, metric_name)): score})
+    if (epoch % self.period) == 0 or self.period == 1:
+      self.run.log_checkpoint(checkpoint_logs)
+      self._latest = None
+    else:
+      self._latest = checkpoint_logs
+
+    return False
+
+  def after_training(self, model):
+    if self._latest is not None:
+      self.run.log_checkpoint(self._latest)
+    return model
+
+
 class XGBRun:
 
-  def __init__(self, params, dtrain, num_boost_round, evals, run_options):
+  def __init__(self, params, dtrain, num_boost_round, evals, verbose_eval, callbacks, run_options):
     self.params = params
     self.dtrain = dtrain
     self.num_boost_round = num_boost_round
     self.evals = evals
-    self.run_options_parsed = parse_run_options(run_options)
+    self.verbose_eval = verbose_eval
+    self.callbacks = callbacks
     self.validation_sets = [(evals, DEFAULT_EVALS_NAME)] if isinstance(evals, DMatrix) else evals
+    self.run_options_parsed = parse_run_options(run_options)
     self.run = None
 
+  def form_callbacks(self):
+    if not self.run_options_parsed['log_checkpoints']:
+      return
+
+    if self.callbacks is None:
+      self.callbacks = []
+    period = MIN_CHECKPOINT_PERIOD
+    if self.callbacks:
+      for cb in self.callbacks:
+        if isinstance(cb, xgboost.callback.EvaluationMonitor):
+          period = cb.period
+    if self.verbose_eval:
+      period = 1 if self.verbose_eval is True else self.verbose_eval
+    period = max(period, (self.num_boost_round + 1) // MAX_NUM_CHECKPOINTS)
```

Comment:
There is an issue to let period as integer, for example num_boost_round = 300,  period = 300 // 200 = 1, all 300 round will call log_checkpoint, which may cause error or lost the latest checkpoints.